Title: Services
Date: 2015-10-06 11:00

## Services

The Informatics group provides bioinformatics analysis support for all users of the Core Sequencing facility and to FAS faculty groups regardless of the origin of their data.    In addition we have the responsibility for sample submission, tracking, scheduling and billing services for many of the core instrument facilities across FAS.

We are small in number and try to cover the needs of as many people across campus as possible.  This means, however, that we can’t provide in-depth or long-term analysis services for single groups.   In these cases we encourage PIs to budget for a bioinformatics post-doc in their grant applications.   We are very happy to provide letters of support and text detailing our expertise,  and outlining possible analyses in grant applications.   Additionally James Cuff in Research Computing is happy to provide letters of support for computational hardware.

### FAS Sequencing Core

Our core sequencing service has two Illumina HiSeqs and a MiSeq.  The Informatics group is committed to providing a fast basecalling and QC service for all users of the facility.   Beyond this we strongly encourage people to get in contact with us (michele_clamp@harvard.edu) and we’ll be happy to discuss things further.  We aim to work alongside researchers with the intention of spreading knowledge and enabling people to run analyses themselves.   As we grow we are building workflows and pipelines to enable more people to be self sufficient in bioinformatics analysis.

In summary we provide:

*   Advice on sequencing needs (length, type and number of reads, recommendations for replicates).
*   Basecalling and demultiplexing of core sequencing runs.  This includes checking the base yield and quality and working to extract data from the remaining reads.
*   Quality control statistics of reads.
*   Advice on which tools are suitable for further analysis.

Other bioinformatics support:

*   Basic unix command help and navigating the cluster
*   How to submit jobs to the queueing system.
*   _De novo_ assembly of transcriptome data.
*   Transcriptome construction both with and without a reference genome.
*   Calculation of expression and differential expression from constructed transcripts.
*   Alignment of related genomes (close and more distant)
*   SNP calling using tools such as GATK/samtools
*   ChIP-Seq analysis mainly using MACS
*   Visualization of NGS data (IGV browser/Tablet  browser/uploading features to UCSC/Ensembl)
*   Annotation of  genes (blast/interpro/pfam/GO)
*   Custom script/algorithm writing.   We have limited resources (although much enthusiasm) for this.   Due to resource limitations at the  moment we try to limit this to areas where we are sure the tools built can benefit multiple groups.
*   Basic R usage and using bioconductor packages for sequence analysis
*   Phylogenetic reontstruction of trees using tools such as phylip, mrbayes, raxml (also GUIs like mega, geneious)
*   Using blast to search common databases.  Constructing custom blast databases.
*   Searching protein domain databases using interproscan.

###  FAS Proteomics Core

We provide  sample submission, tracking and result delivery.    We can also provide access to common proteomics analysis tools such as ProteomeDiscoverer, MaxQuant and TPP.    We’re also happy to install other software as required.     For expert advice on analysis techniques we encourage people to contact Bogdan Budnik,  Sunia Trauger or Bill Lane in the Proteomics Core.
